### PR TITLE
Issue #1011 Fix Adding developer to sudoers when using OpenShift 3.6.x

### LINKED ISSUE #1011: Adding developer to sudoers fails when using 3.6.x version of OpenShift
Looks like we have some code regression the way `minishift_kubeconfig` created and this make minishift failed during adding `developer` user to sudoers.

@jorgemoralespou 

```
$ ./minishift start --openshift-version v3.6.0-alpha.2
[...]
Image pull complete
OpenShift server started.

The server is accessible via web console at:
    https://192.168.42.71:8443

You are logged in as:
    User:     developer

To login as administrator:
    oc login -u system:admin

Error during post cluster up configuration: Unable to add sudoer role
```

### PR DESCRIPTION
**Pull request by praveenkumar**
Fix #1011

### CODE REVIEW

**Comment by budhrg** [comment]
@praveenkumar , you mean by `3.6.*` instead of `1.3.6` in commit message right? Need update if so.

**Comment by budhrg** [comment]
This PR is working as expected for me with https://ci.appveyor.com/api/buildjobs/0w0waa3d1w8mfgvn/artifacts/out%2Fwindows-amd64%2Fminishift.exe. 

@hferentschik , please integrate this with https://github.com/minishift/minishift/pull/1017 before merging it and see if the issue still exists or not.
For me, I am still getting same error when cherry-picking this PR commit into your PR.

**Comment by hferentschik** [comment]
> please integrate this with #1017 before merging it and see if the issue still exists or not.

Why? The problem is fully reproducible and testable w/o #1017.

File: pkg/minishift/clusterup/clusterup.go [hunk 1]
```diff
@@ -75,7 +77,11 @@ func ClusterUp(config *ClusterUpConfig, clusterUpParams map[string]string, runne
 
 // PostClusterUp runs the Minishift specific provisioning after 'cluster up' has run
 func PostClusterUp(clusterUpConfig *ClusterUpConfig, sshCommander provision.SSHCommander, addOnManager *manager.AddOnManager) error {
-	err := kubeconfig.CacheSystemAdminEntries(clusterUpConfig.KubeConfigPath, getConfigClusterName(clusterUpConfig.Ip, clusterUpConfig.Port))
+	isVersion3_6, err := ValidateOpenshiftMinVersion(clusterUpConfig.OpenShiftVersion, "v3.6.0-alpha.0")
```
Comment (hferentschik) [inline pkg/minishift/clusterup/clusterup.go]: will this work with ocp?

File: pkg/minikube/kubeconfig/config.go [hunk 2]
```diff
@@ -59,7 +60,7 @@ func GetConfigPath() string {
 }
 
 // Cache system admin entries to be used to run oc commands
-func CacheSystemAdminEntries(systemEntriesConfigPath, clusterName string) error {
+func CacheSystemAdminEntries(systemEntriesConfigPath, clusterName string, isVersion3_6 bool) error {
```
Comment (hferentschik) [inline pkg/minikube/kubeconfig/config.go]: I don't get why you have to pass _isVersion3_6_ into this method. IMO you can keep this method as is and just change the clusterName parameter clusterup.go. Then the need to check the version stays in one spot.

**Comment by hferentschik** [comment]
> This PR is working as expected for me 

How did you verify this? For me it still does not work as expected. Here is what I did.

* `rm ~/.kube/config`
* `minishift start --vm-driver xhyve` # OK will provision 1.5.1
* `minishift delete`
* `minishift start --vm-driver xhyve --openshift-version v3.6.0-alpha.2` # OK will provision 3.6.0-alpha.2
* `minishift delete`
* `minishift start --vm-driver xhyve` # FAIL "Error during post cluster up configuration: Unable to add sudoer role"

@budhrg, @praveenkumar - Is the above sequence working for you?

**Comment by budhrg** [comment]
@hferentschik , looks like someone reproduce it here https://github.com/minishift/minishift/issues/1053.

**Comment by hferentschik** [comment]
I have been over this - https://github.com/minishift/minishift/pull/1041#issuecomment-309879842 - multiple times and I keep getting the same behaviour. Even when I delete the full ~/.kube directory. 

Once it fails, I'll delete ~/.kube/config and start again and it works. 

At that stage I will need to add more debug trace to see what the exact error is.

**Comment by budhrg** [comment]
@LalatenduMohanty @praveenkumar , can one of you try this PR in both Mac or Linux ? 
Well I don't want to say "it can be OS specific" but let's confirm it too. Also, James seems to have reproduce it here  https://github.com/minishift/minishift/pull/1041 in Mac. 

I am not able to reproduce in Linux (F25).
Sequences to test : 
* `rm ~/.kube/config`
* `minishift start`
* `minishift delete`
* `minishift start --openshift-version v3.6.0-alpha.2`
* `minishift delete`
* `minishift start`

File: pkg/minikube/kubeconfig/config.go [hunk 3]
```diff
@@ -81,8 +81,10 @@ func CacheSystemAdminEntries(systemEntriesConfigPath, clusterName string) error
 			break
 		}
 	}
-
 	userName := fmt.Sprintf("system:admin/%s", clusterName)
+	if isVersion3_6 {
+		userName = "system:admin/127-0-0-1:8443"
+	}
 	for k, v := range config.Users {
 		if v.Name == userName || v.Name == "system:admin/127-0-0-1:8443" {
```
Comment (hferentschik) [inline pkg/minikube/kubeconfig/config.go]: @praveenkumar @budhrg there bug in the code is here: 

    if v.Name == userName || v.Name == "system:admin/127-0-0-1:8443"

To start with the _v.Name == "system:admin/127-0-0-1:8443_ is generally not needed anymore with the introduction of the _isVersion3_6_ flag. If __isVersion3_6_ is true, both condition are actually identical. However, what is worse, it makes things fail independent of the OpenShift version. config.Users is iterating over a map which does not guarantee any order. Remember the 127-0-0-1:8443 is not gone form the config. It seems that on Linux the iteration order is different than on OS X. Removing _|| v.Name == "system:admin/127-0-0-1:8443"_ should do the trick.

File: pkg/minikube/kubeconfig/config.go [hunk 4]
```diff
@@ -60,7 +60,7 @@ func GetConfigPath() string {
 }
 
 // Cache system admin entries to be used to run oc commands
-func CacheSystemAdminEntries(systemEntriesConfigPath, clusterName string) error {
+func CacheSystemAdminEntries(systemEntriesConfigPath, clusterName string, isVersion3_6 bool) error {
```
Comment (hferentschik) [inline pkg/minikube/kubeconfig/config.go]: maybe _isOpenShift3_6_ as flag name.

File: pkg/minikube/kubeconfig/config.go (hunk 2 above)
Comment (LalatenduMohanty) [inline pkg/minikube/kubeconfig/config.go]: I am little confused. The change in logic for all versions of Origin > = 3.6 , right? But the code seems to be on only 3.6 version..

**Comment by LalatenduMohanty** [comment]
@budhrg @praveenkumar It is failing for me 

```
-- Creating initial project "myproject" ... OK
-- Removing temporary directory ... OK
-- Checking container networking ... OK
-- Server Information ... 
   OpenShift server started.
   The server is accessible via web console at:
       https://192.168.64.7:8443

   You are logged in as:
       User:     developer
       Password: developer

   To login as administrator:
       oc login -u system:admin

Error during post cluster up configuration: Unable to add sudoer role
```

**Comment by budhrg** [comment]
> It is failing for me on macOS.

and in Fedora ?

File: pkg/minikube/kubeconfig/config.go (hunk 3 above)
Comment (praveenkumar) [inline pkg/minikube/kubeconfig/config.go]: @hferentschik looks like I made a mistake remove that from my current PR.

**Comment by budhrg** [comment]
@LalatenduMohanty Is the explanation of bug https://github.com/minishift/minishift/pull/1041#discussion_r123334995.

The reason I was not able to reproduce is I had the previous commit where I didn't have the follow code snippet.
```
v.Name == "system:admin/127-0-0-1:8443"
```

cc @hferentschik

**Comment by hferentschik** [comment]
> The reason I was not able to reproduce is I had the previous commit where I didn't have the follow code snippet.

then I have no idea what you tested?!

Either way, it looks like the code is updated.

File: pkg/minikube/kubeconfig/config.go [hunk 5]
```diff
@@ -60,7 +60,7 @@ func GetConfigPath() string {
 }
 
 // Cache system admin entries to be used to run oc commands
-func CacheSystemAdminEntries(systemEntriesConfigPath, clusterName string) error {
+func CacheSystemAdminEntries(systemEntriesConfigPath, clusterName string, isOpenshift3_6 bool) error {
```
Comment (hferentschik) [inline pkg/minikube/kubeconfig/config.go]: I am still not happy about the fact that the version check "leaks" even into _CacheSystemAdminEntries_. Here is a proposal for a change - https://github.com/praveenkumar/minishift/pull/1

**Comment by hferentschik** [comment]
Merged, thanks.

**Pull request merged**